[AKS] Fix S3 access issue in AKS

gangsf opened the pull request:
## Why are these changes needed?

We met S3 access problem during the AKS test 

```
Failed to update Serve deployment ObjectDetection. Reason: Replica startup failure
The deployment failed to start 3 times in a row. This may be due to a problem with its constructor or initial health check failing. See controller logs for details. Error:
ray::ServeReplica:default:ObjectDetection.initialize_and_get_metadata() (pid=689, ip=10.0.192.83, actor_id=486dcad012ce012b7648f9bf01000000, repr=<ray.serve._private.replica.ServeReplica:default:ObjectDetection object at 0x7f23ebfcf280>)
  File "/home/ray/anaconda3/lib/python3.9/concurrent/futures/_base.py", line 439, in result
    return self.__get_result()
  File "/home/ray/anaconda3/lib/python3.9/concurrent/futures/_base.py", line 391, in __get_result
    raise self._exception
  File "/home/ray/anaconda3/lib/python3.9/site-packages/ray/serve/_private/replica.py", line 1058, in initialize_and_get_metadata
    await self._replica_impl.initialize(deployment_config)
  File "/home/ray/anaconda3/lib/python3.9/site-packages/ray/serve/_private/replica.py", line 761, in initialize
    raise RuntimeError(traceback.format_exc()) from None
RuntimeError: Traceback (most recent call last):
  File "/home/ray/anaconda3/lib/python3.9/site-packages/ray/serve/_private/replica.py", line 739, in initialize
    await self._user_callable_wrapper.initialize_callable()
  File "/home/ray/anaconda3/lib/python3.9/site-packages/ray/serve/_private/replica.py", line 1477, in initialize_callable
    await self._call_func_or_gen(
  File "/home/ray/anaconda3/lib/python3.9/site-packages/ray/serve/_private/replica.py", line 1421, in _call_func_or_gen
    result = callable(*args, **kwargs)
  File "/tmp/ray/session_2025-09-08_14-22-29_126045_77/runtime_resources/working_dir_files/abfss_dataplane-staging-blob_anyscalecloudstaging_dfs_core_windows_net_org_7c1Kalm9WcX2bNIjW53GUT_cld_5nnv7pt2jn2312x2e5v72z53n2_runtime_env_packages_pkg_4b51a6325a53213d5453c8186d6f3543/object_detection.py", line 74, in __init__
    self.model = self._load_faster_rcnn_model()
  File "/tmp/ray/session_2025-09-08_14-22-29_126045_77/runtime_resources/working_dir_files/abfss_dataplane-staging-blob_anyscalecloudstaging_dfs_core_windows_net_org_7c1Kalm9WcX2bNIjW53GUT_cld_5nnv7pt2jn2312x2e5v72z53n2_runtime_env_packages_pkg_4b51a6325a53213d5453c8186d6f3543/object_detection.py", line 82, in _load_faster_rcnn_model
    with smart_open(REMOTE_MODEL_PATH, "rb") as s3_file:
  File "/home/ray/anaconda3/lib/python3.9/site-packages/smart_open/smart_open_lib.py", line 224, in open
    binary = _open_binary_stream(uri, binary_mode, transport_params)
  File "/home/ray/anaconda3/lib/python3.9/site-packages/smart_open/smart_open_lib.py", line 400, in _open_binary_stream
    fobj = submodule.open_uri(uri, mode, transport_params)
  File "/home/ray/anaconda3/lib/python3.9/site-packages/smart_open/s3.py", line 224, in open_uri
    return open(parsed_uri['bucket_id'], parsed_uri['key_id'], mode, **kwargs)
  File "/home/ray/anaconda3/lib/python3.9/site-packages/smart_open/s3.py", line 291, in open
    fileobj = Reader(
  File "/home/ray/anaconda3/lib/python3.9/site-packages/smart_open/s3.py", line 574, in __init__
    self.seek(0)
  File "/home/ray/anaconda3/lib/python3.9/site-packages/smart_open/s3.py", line 666, in seek
    self._current_pos = self._raw_reader.seek(offset, whence)
  File "/home/ray/anaconda3/lib/python3.9/site-packages/smart_open/s3.py", line 417, in seek
    self._open_body(start, stop)
  File "/home/ray/anaconda3/lib/python3.9/site-packages/smart_open/s3.py", line 438, in _open_body
    response = _get(
  File "/home/ray/anaconda3/lib/python3.9/site-packages/smart_open/s3.py", line 330, in _get
    return client.get_object(Bucket=bucket, Key=key, Range=range_string)
  File "/home/ray/anaconda3/lib/python3.9/site-packages/botocore/client.py", line 535, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/home/ray/anaconda3/lib/python3.9/site-packages/botocore/client.py", line 963, in _make_api_call
    http, parsed_response = self._make_request(
  File "/home/ray/anaconda3/lib/python3.9/site-packages/botocore/client.py", line 986, in _make_request
    return self._endpoint.make_request(operation_model, request_dict)
  File "/home/ray/anaconda3/lib/python3.9/site-packages/botocore/endpoint.py", line 119, in make_request
    return self._send_request(request_dict, operation_model)
  File "/home/ray/anaconda3/lib/python3.9/site-packages/botocore/endpoint.py", line 198, in _send_request
    request = self.create_request(request_dict, operation_model)
  File "/home/ray/anaconda3/lib/python3.9/site-packages/botocore/endpoint.py", line 134, in create_request
    self._event_emitter.emit(
  File "/home/ray/anaconda3/lib/python3.9/site-packages/botocore/hooks.py", line 412, in emit
    return self._emitter.emit(aliased_event_name, **kwargs)
  File "/home/ray/anaconda3/lib/python3.9/site-packages/botocore/hooks.py", line 256, in emit
    return self._emit(event_name, kwargs)
  File "/home/ray/anaconda3/lib/python3.9/site-packages/botocore/hooks.py", line 239, in _emit
    response = handler(**kwargs)
  File "/home/ray/anaconda3/lib/python3.9/site-packages/botocore/signers.py", line 105, in handler
    return self.sign(operation_name, request)
  File "/home/ray/anaconda3/lib/python3.9/site-packages/botocore/signers.py", line 189, in sign
    auth.add_auth(request)
  File "/home/ray/anaconda3/lib/python3.9/site-packages/botocore/auth.py", line 418, in add_auth
    raise NoCredentialsError()
botocore.exceptions.NoCredentialsError: Unable to locate credentials
```

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

We verified in Anyscale env. It works end to end.  https://console.anyscale-staging.com/cld_5nnv7pt2jn2312x2e5v72z53n2/prj_y8syktydl7ltabhz5axdelwnce/services/service2_8rpblixketpcvmd8is8md62ssk?service-tabs=overview



<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Use a boto3 client (signed if creds exist, unsigned otherwise) with smart_open to stream the model from S3 to cluster storage, updating both the Serve app and eval notebook.
> 
> - **Object detection service (`doc/source/ray-overview/examples/object-detection/object_detection.py`)**
>   - Use `boto3` session to select signed or unsigned client based on available credentials and pass via `transport_params` to `smart_open`.
>   - Ensure target directory exists and stream model download in chunks to `CLUSTER_MODEL_PATH`.
> - **Evaluation notebook (`2.object_detection_batch_inference_eval.ipynb`)**
>   - Add `boto3`-based S3 client with unsigned fallback and use `transport_params` for `smart_open` when downloading the model to cluster storage.
>   - Minor text tweak in Anyscale configuration note.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit ea2db7553d53a45b258b0e3e99ee5a723a31e1ad. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

Commit:
[AKS] Fix S3 access issue in AKS

Signed-off-by: Gang Zhao <gang@gang-JQ62HD2C37.local>

kunling-anyscale reviewed (approved):
tested the notebook with new code, works well

Commit:
[AKS] Use boto3 credential fallback for S3 access

- Follow the same pattern as python/ray/_private/runtime_env/protocol.py
- Check for credentials first, fall back to unsigned client for public buckets
- Remove URL conversion approach in favor of native S3 URI handling
- Apply to both object_detection.py and batch inference notebook
- This allows accessing public S3 buckets without AWS credentials

Signed-off-by: Gang Zhao <gang@gang-JQ62HD2C37.local>

gangsf commented on doc/source/ray-overview/examples/object-detection/object_detection.py:
```diff
@@ -78,10 +83,47 @@ def __init__(self):
+        #     s3 = boto3.client("s3", config=Config(signature_version=UNSIGNED))  # <-- sentinel, not the string "unsigned"
+        #     with smart_open(REMOTE_MODEL_PATH, "rb", transport_params={"client": s3}) as src, open(CLUSTER_MODEL_PATH, "wb") as dst:
+        #         for chunk in iter(lambda: src.read(1024 * 1024), b""):
+        #             dst.write(chunk)
```
Removed

jjyao commented on doc/source/ray-overview/examples/object-detection/object_detection.py:
```diff
@@ -78,10 +83,33 @@ def __init__(self):
         """Loads the Faster R-CNN model from a remote source if not already available locally."""
         # Download model only once from the remote storage to the cluster path.
+
+        def to_https_from_s3(uri: str) -> str:
+            """
+            Convert s3://bucket/key (or s3a://, s3n://) to
+            https://bucket.s3.amazonaws.com/key. If already http(s), return as-is.
+            """
+            u = urlparse(uri)
+            if u.scheme in ("http", "https"):
+                return uri
+            if u.scheme in ("s3", "s3a", "s3n"):
+                bucket = u.netloc
+                key = u.path.lstrip("/")
+                if not bucket or not key:
+                    raise ValueError(f"Invalid S3 URI: {uri}")
+                return f"https://{bucket}.s3.amazonaws.com/{key}"
+            raise ValueError(f"Unsupported URI scheme in REMOTE_MODEL_PATH: {uri}")
+
+        HTTPS_URL = to_https_from_s3(REMOTE_MODEL_PATH)
```
why do we do this? can we just use s3 client instead?

gangsf replied:
Public S3 bucket: The model is stored in a publicly accessible S3 bucket
HTTPS access: https://bucket.s3.amazonaws.com/key format allows direct HTTP access
No authentication: Bypasses the need for AWS credentials or boto3 configuration

Commit:
Fix curly apostrophes in notebook

- Replace curly apostrophes (') with straight apostrophes (')
- Fixes you'll, won't, and Anyscale's
- Ensures consistent character encoding

Signed-off-by: Gang Zhao <gang@gang-JQ62HD2C37.local>

gangsf committed:
Merge branch 'master' into gang/fix_s3_access_in_AKS

gangsf committed:
Merge branch 'master' into gang/fix_s3_access_in_AKS
